### DISCUSSION TITLE
Horizon - Fill Screen

### DISCUSSION THREAD

**Discussion by r0w4n** [XCSoar/XCSoar#1222] [Polls]
Hello,

I was thinking about changing the horizon so that is not bound by a circle, as it currently is. Instead the horizon would take up the all of the draw space available.

Current:
<img width="515" alt="Screenshot 2023-05-18 at 08 27 18" src="https://github.com/XCSoar/XCSoar/assets/6553343/149dc022-32bf-4b6d-b571-e628fd9438c0">

Suggested::
![new-Screenshot 2023-05-18 at 08 27 18](https://github.com/XCSoar/XCSoar/assets/6553343/d4c84d54-4ddd-4ea7-b314-fa667457af6a)

I was having a play in C++ as I was just playing with the existing [HorizonRenderer.cpp](https://github.com/XCSoar/XCSoar/blob/master/src/Renderer/HorizonRenderer.cpp) but then I thought, maybe this could be done in Lua instead?

So before I start properly, I wanted to ask here the following

**Comment by MaxKellermann** [comment]
How do you want to "do it in Lua"? Do you want to expose Canvas methods to Lua, and call into Lua while rendering the screen?
That would be a bad idea; Lua can be used for high-level code, for controlling the UI, but not for rendering something. The performance of the renderer is critical, and that should be C++.

**Reply by r0w4n** [reply]
Thanks @MaxKellermann that's really good to know. I'll explore this in C++ then.

**Comment by lordfolken** [comment]
While my wish for a Garmin style synthetic vision display may be a bit far fetched...

What is really needed:
* bank angle indicator
* pitch angle indicator

Nice to have
* track with waypoint navigation
* altitude tape with VSI
* speed tape with mc speed / stall speed / vne speed and in circling mode optimal thermalling speed
* optimal bank angle indicator during thermalling

**Reply by r0w4n** [reply]
That's a nice wish list. I really like the last two.

**Comment by r0w4n** [comment]
I've been working at providing a full screen horizon. It has proved to be quite complex primarily due to an assumption which is that I have to keep all of the vectors inside the render area. The reason for this assumption is that if you have the horizon in an info box, any overflow is seen in the the other info boxes or main page elements e.g.: 

<img width="406" alt="Screenshot 2023-05-27 at 12 43 22" src="https://github.com/XCSoar/XCSoar/assets/6553343/69520a17-eaf1-4fac-8155-cd9cd5de3764">

I wanted to ask here whether:
1. My assumption is correct, that all vertices need to be in the render area? As I have said this makes it more complex.
2. Whether there's a way to specify the view port of the info box so that it doesn't overflow into other info boxes or the main page?

**Reply by r0w4n** [reply]
edited to provide the correct image

**Reply by MaxKellermann** [reply]
Depends on the operating system. OpenGL does no clipping by default, and you can draw over the current `Canvas` boundaries; but you can use `GLCanvasScissor` to enable clipping. This adds some overhead and thus slows down rendering.

**Comment by lordfolken** [comment]
There is a PR open for a much better looking artificial horizon:  https://github.com/XCSoar/XCSoar/pull/1266 
If someone would care to solve the detailing...
